IdentityCore/src/telemetry/MSIDDefaultDispatcher.m

```diff
@@ -0,0 +1,136 @@
+    if (self)
+    {
+        _eventsToBeDispatched = [NSMutableDictionary new];
+        _errorEvents = [NSMutableSet new];
+        _observer = observer;
+        _setTelemetryOnFailure = setTelemetryOnFailure;
+        NSString *queueName = [NSString stringWithFormat:@"com.microsoft.dispatcher-%@", [NSUUID UUID].UUIDString];
+        _synchronizationQueue = dispatch_queue_create([queueName cStringUsingEncoding:NSASCIIStringEncoding], DISPATCH_QUEUE_SERIAL);
+    }
+    return self;
+}
+
+- (BOOL)containsObserver:(id<MSIDTelemetryEventsObserving>)dispatcher
+{
+    return self.observer == dispatcher;
+}
+
+- (void)flush:(NSString *)requestId
+{
+    NSArray<id<MSIDTelemetryEventInterface>> *events = [self popEventsForReuquestId:requestId];
+    
+    if ([events count])
+    {
+        __auto_type defaultEvent = [[MSIDTelemetryDefaultEvent alloc] initWithName:MSID_TELEMETRY_EVENT_DEFAULT_EVENT context:nil];
+        NSMutableArray *eventsToBeDispatched = [@[[defaultEvent getProperties]] mutableCopy];
+        
+        for (id<MSIDTelemetryEventInterface> event in events)
+        {
+            [eventsToBeDispatched addObject:[event getProperties]];
+        }
+        
+        [self dispatchEvents:eventsToBeDispatched];
+    }
+}
+
+- (void)receive:(NSString *)requestId event:(id<MSIDTelemetryEventInterface>)event
+{
+    if ([NSString msidIsStringNilOrBlank:requestId] || !event) return;
+    
+    dispatch_sync(self.synchronizationQueue, ^{
+        NSMutableArray *eventsForRequestId = self.eventsToBeDispatched[requestId];
+        if (!eventsForRequestId)
+        {
+            eventsForRequestId = [NSMutableArray new];
+            [self.eventsToBeDispatched setObject:eventsForRequestId forKey:requestId];
+        }
+        
+        [eventsForRequestId addObject:event];
+        
+        if (event.errorInEvent) [self.errorEvents addObject:requestId];
+    });
+}
+
+- (void)dispatchEvents:(NSArray<NSDictionary<NSString *, NSString *> *> *)rawEvents
+{
+    NSMutableArray *eventsToBeDispatched = [NSMutableArray new];
+    
+    for (NSDictionary *event in rawEvents)
+    {
+        [eventsToBeDispatched addObject:[self appendPrefixForEvent:event]];
+    }
+    
+    [self.observer onEventsReceived:eventsToBeDispatched];
+}
+
+#pragma mark - Protected
+
+- (NSArray *)popEventsForReuquestId:(NSString *)requestId
+{
+    __block NSArray *events;
+    dispatch_sync(self.synchronizationQueue, ^{
+        BOOL errorInEvent = [self.errorEvents containsObject:requestId];
+        
+        // Remove requestId as we won't need it anymore
+        [self.errorEvents removeObject:requestId];
+        
+        if (self.setTelemetryOnFailure && !errorInEvent) return;
+        
+        events = [self.eventsToBeDispatched[requestId] copy];
+        [self.eventsToBeDispatched removeObjectForKey:requestId];
+    });
+    
+    return events;
+}
+
+#pragma mark - Private
+
+- (NSDictionary *)appendPrefixForEvent:(NSDictionary *)event
```

Comment:
just to double check: this applies also for aggregated dispatcher?

Reply:
yes